File: 394. Decode String/decodeString.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        num = 0
        res = ""
        for c in s:
            if c >= '0' and c <= '9':
                num = num * 10 + int(c)
            elif c == '[':
                stack.append((res, num))
                num = 0
                res = ""
            elif c == ']':
                top = stack.pop()
                res = top[0] + res * top[1]
            else:
                res += c
        return res

File: 394. Decode String/decodeString.java
class Solution {
    public String decodeString(String s) {
        return dfs(s, 0)[0];
    }

    public String[] dfs(String s, int p) {
        StringBuilder res = new StringBuilder();
        int num = 0;
        while (p < s.length()) {
            char curr = s.charAt(p);
            if (curr >= '0' && curr <= '9') {
                num = num * 10 + (curr - '0');
            } else if (curr == '[') {
                String[] temp = dfs(s, p + 1);
                p = Integer.parseInt(temp[0]);
                while (num > 0) {
                    res.append(temp[1]);
                    num--;
                }
            } else if (curr == ']') {
                return new String[] {String.valueOf(p), res.toString()};
            } else {
                res.append(String.valueOf(curr));
            }
            p++;
        }
        return new String[] {res.toString()};
    }
}